src/core/xfa/xfa_object.js

```diff
@@ -0,0 +1,253 @@
+/* Copyright 2021 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { getInteger, getKeyword } from "./utils.js";
+import { warn } from "../../shared/util.js";
+
+// We use these symbols to avoid name conflict between tags
+// and properties/methods names.
+const $clean = Symbol();
+const $cleanup = Symbol();
+const $content = Symbol("content");
+const $dump = Symbol();
+const $finalize = Symbol();
+const $isTransparent = Symbol();
+const $namespaceId = Symbol("namespaceId");
+const $nodeName = Symbol("nodeName");
+const $onChild = Symbol();
+const $onChildCheck = Symbol();
+const $onText = Symbol();
+
+const _defaultValue = Symbol();
+const _hasChildren = Symbol();
+const _options = Symbol();
+const _validator = Symbol();
+
+class XFAObject {
+  constructor(nsId, name, hasChildren = false) {
+    this[$namespaceId] = nsId;
+    this[$nodeName] = name;
+    this[_hasChildren] = hasChildren;
+  }
+
+  [$onChild](child) {
+    if (!this[_hasChildren] || !this[$onChildCheck](child)) {
+      return;
+    }
+
+    const name = child[$nodeName];
+    const node = this[name];
+    if (Array.isArray(node)) {
+      node.push(child);
+    } else if (node === null) {
+      this[name] = child;
+    } else {
+      warn(`XFA - node "${this[$nodeName]}" accepts only one child: ${name}`);
+    }
+  }
+
+  [$onChildCheck](child) {
+    return (
+      this.hasOwnProperty(child[$nodeName]) &&
+      child[$namespaceId] === this[$namespaceId]
+    );
+  }
+
+  [$onText](_) {}
+
+  [$finalize]() {}
+
+  [$clean](builder) {
+    delete this[_hasChildren];
+    if (this[$cleanup]) {
+      builder.clean(this[$cleanup]);
+      delete this[$cleanup];
+    }
+  }
+
+  [$isTransparent]() {
+    return this.name === "";
+  }
+
+  [$dump](buf = [], level = 0) {
```

Comment:
Is there an advantage to using our own output format vs json stringifying an object?

Reply:
I just discover `toJSON`: I'll update the patch with.
Thx for the tip.